vespalib/src/vespa/vespalib/util/signalhandler.cpp

```diff
@@ -116,23 +153,78 @@ SignalHandler::unhook()
     sigaction(_signal, &act, nullptr);
 }
 
+void
+SignalHandler::enable_cross_thread_stack_tracing()
+{
+    USR2.hook(); // Our secret sauce
+    _shared_backtrace_data._signal_is_hooked = true;
+}
+
+/*
+ * There's no existing API for conveniently getting the stack trace of any other thread
+ * than your own, so we have to turn things around a bit and make the target thread
+ * dump its own stack in a way that we can then safely access and return it.
+ *
+ * We allow for hooking SIGUSR2 at process startup time and use this as our dedicated
+ * stack dump signal internally. The process of stack dumping is then basically:
+ *
+ *  1. Caller sends SIGUSR2 to the target thread. This happens within a global mutex
+ *     that ensures no other callers can attempt to dump stack at the same time. This
+ *     is because we use a single, globally shared state between caller and target
+ *     (the same signal handler function is used by all threads).
+ *  2. Caller acquire-polls (with 1ms sleep) for target thread signal handler completion.
+ *     Fancy technologies such as mutexes and condition variables are not safe to use
+ *     from within a signal handler and therefore cannot be used.
+ *  3. Target thread suddenly finds itself in Narnia (signal handler). Since the
+ *     signal is the magical SIGUSR2, it proceeds to dump its stack frame addresses in
+ *     a shared buffer. It then toggles completion, with release semantics, before
+ *     returning to its regularly scheduled programming.
+ *  4. Caller exits poll-loop and assembles a complete stack trace from the frame
+ *     addresses in the shared buffer, all demangled and shiny.
+ */
+string
+SignalHandler::get_cross_thread_stack_trace(pthread_t thread_id)
+{
+    if (!_shared_backtrace_data._signal_is_hooked) {
+        return "(cross-thread stack tracing is not enabled in process)";
+    }
+    // This will only work with pthreads, but then again, so will Vespa.
+    if (thread_id == pthread_self()) {
+        return vespalib::getStackTrace(3); // Skip getStackTrace() internals + this function's frame.
+    }
+
+    static std::mutex stack_dump_caller_mutex;
+    std::lock_guard guard(stack_dump_caller_mutex);
+
+    assert(!_shared_backtrace_data._want_backtrace.load());
+    _shared_backtrace_data._signal_handler_done.store(false);
+
+    _shared_backtrace_data._want_backtrace.store(true);
+    if (pthread_kill(thread_id, SIGUSR2) != 0) {
+        _shared_backtrace_data._want_backtrace.store(false);
+        return "(pthread_kill() failed; could not get backtrace)";
+    }
+    while (!_shared_backtrace_data._signal_handler_done.load()) {
```

Comment:
consider using cmp exchange here as well to consume done signal

Reply:
Done